### DISCUSSION TITLE
[General] Can we impersonate wallet accounts using era_test_node?

### DISCUSSION THREAD

**Discussion by deadshotryker** [zkSync-Community-Hub/zksync-developers#272] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

We are trying to simulate a transaction where we impersonate a smart contract (Gnosis safe) to execute transactions on a mainnet fork. 

Our current smart contracts hold about 6mn$ of user funds but we cannot deploy upgrades to it without the ability to simulate the changes we will make.

### What is the specific issue or error you're encountering?

Can't figure out how to impersonate mainnet wallets. 

### Can you share the error messages or logs you're receiving, if any?

NA

### Have you made any recent changes to the contract before encountering this issue?

NA

### Are there any external libraries or contracts that your contract interacts with?

https://github.com/matter-labs/era-test-node

### Can you provide the relevant portions of your contract code where the issue is occurring?

NA

### Have you tried to isolate the problem, and if so, what were the results?

Couldn't find the documentation for this so could not figure out how to begin.

### What steps have you already taken to try to resolve the issue?

Reached out to the team and dev rel team for assistance.

### Additional Details

Would be awesome if we can a very simple way to simulate tx's with the ability to change the sender. Like tenderly.

**Comment by MexicanAce** [comment, marked as answer]
Hi @deadshotryker 👋 !

I think you're looking for the [hardhat_impersonateAccount](https://github.com/matter-labs/era-test-node/blob/main/SUPPORTED_APIS.md#hardhat_impersonateaccount) endpoint.

See here for an example of how we use it in our tests : https://github.com/matter-labs/era-test-node/blob/main/e2e-tests/test/hardhat-apis.test.ts#L67-L90

**Comment by procebetjumorales** [comment]
Simulating transactions with the ability to change the sender (impersonate wallets) on a mainnet fork can be achieved using the era-test-node tool provided by Matter Labs.

 This tool allows for easy simulation of transactions, providing a simple and efficient way to test upgrades and changes to smart contracts.

 For further assistance, reach out to the development team or dev rel team for guidance on utilizing era-test-node effectively.

Happy testing and upgrading your smart contracts!

**Comment by dgeyston** [comment]
To simulate transactions on a mainnet fork and impersonate smart contracts like Gnosis Safe, you can utilize tools like Hardhat with a mainnet fork plugin or Ganache with mainnet forking capabilities. These tools allow you to interact with contracts on a forked version of the Ethereum mainnet, enabling simulation of transactions and upgrades before deployment.

**Comment by mariyokaveen** [comment]
To simulate transactions with the ability to impersonate mainnet wallets, you may consider using test node environments like the era-test-node provided by Matter Labs. Alternatively, explore tools like Tenderly which offer simulation features with sender customization.

**Comment by m4ccmocare** [comment]
The user is attempting to simulate a transaction where they impersonate a smart contract (Gnosis safe) to execute transactions on a mainnet fork. Their current smart contracts hold about $6 million of user funds, but they cannot deploy upgrades without the ability to simulate the changes. They are encountering issues with figuring out how to impersonate mainnet wallets and have not found relevant documentation. They have reached out to the team and dev rel team for assistance. The user is looking for a simple way to simulate transactions with the ability to change the sender, similar to Tenderly. No repository link was provided.
